internal/server/multitherading.go

```diff
@@ -0,0 +1,108 @@
+package server
+
+import (
+	"bytes"
+
+	"github.com/dicedb/dice/internal/cmd"
+	"github.com/dicedb/dice/internal/comm"
+	"github.com/dicedb/dice/internal/eval"
+	"github.com/dicedb/dice/internal/ops"
+	"github.com/dicedb/dice/internal/shard"
+	"github.com/twmb/murmur3"
+)
+
+func getShard(key string, n uint32) uint32 {
+	hash := murmur3.Sum32([]byte(key))
+	return hash % n
+}
+
+func (s *AsyncServer) cmdsBreakup(redisCmd *cmd.RedisCmd, c *comm.Client) []cmd.RedisCmd {
+	val, ok := WorkerCmdsMeta[redisCmd.Cmd]
+
+	if !ok {
+		return []cmd.RedisCmd{*redisCmd}
+	}
+
+	// if command supports multisharding then send the command
+	// to the respective breakup function
+	// Which can return array of broken down commands
+	return val.Breakup(s.shardManager, redisCmd, c)
+}
+
+func (s *AsyncServer) scatter(cmds []cmd.RedisCmd, c *comm.Client) {
+	// single sharded command
+	if len(cmds) == 1 {
+		var id uint32
+		if len(cmds[0].Args) > 0 {
+			key := cmds[0].Args[0]
+			id = getShard(key, uint32(s.shardManager.GetShardCount()))
+		}
+		s.shardManager.GetShard(shard.ShardID(id)).ReqChan <- &ops.StoreOp{
+			Cmd:      &cmds[0],
+			WorkerID: "server",
+			ShardID:  int(id),
+			Client:   c,
+		}
+	} else {
+		// multishard command
+		// Condition for command that requires all shards such as PING
+		if len(cmds) == s.shardManager.GetShardCount() {
```

Comment:
The if and else need to be handled the same way. 

It should not be anything based this logic - `len(cmds) == s.shardManager.GetShardCount()`.
If I have a command MSET k1, v1, k2, v2, k3, v3, then I just need to scatter it based on the keys. Same should happen if I get SET k1, v1. It shouldn't matter how many shards I have.

Reply:
Thanks for this. Removed. (Actually that condition is to send any command to all shards, but that doesn't make sense.